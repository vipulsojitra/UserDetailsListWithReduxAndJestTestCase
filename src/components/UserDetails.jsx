import React, { useEffect, useState } from "react";
import { usersList, usersDelete, usersUpdate } from "../Api";
import { connect } from "react-redux";

import Button from "@mui/material/Button";
import TextField from "@mui/material/TextField";
import Dialog from "@mui/material/Dialog";
import DialogActions from "@mui/material/DialogActions";
import DialogContent from "@mui/material/DialogContent";
import DialogTitle from "@mui/material/DialogTitle";
import Table from "@mui/material/Table";
import TableBody from "@mui/material/TableBody";
import TableCell from "@mui/material/TableCell";
import TableContainer from "@mui/material/TableContainer";
import TableHead from "@mui/material/TableHead";
import TableRow from "@mui/material/TableRow";
import Paper from "@mui/material/Paper";
import { Grid } from "@mui/material";

export function UserDetails({
  usersList,
  usersDelete,
  usersUpdate,
  Data: { userList },
}) {
  const [open, setOpen] = useState(false);
  const [userData, setUserData] = useState({});
  useEffect(() => {
    usersList();
  }, []);

  const handleEdit = (id) => {
    userList.map((item) => {
      if (item.id == id) {
        setUserData(item);
      }
    });
    setOpen(true);
  };

  const handleUpdate = () => {
    usersUpdate(userData);
    setOpen(false);
  };
  const handleClose = () => {
    setOpen(false);
  };

  const handleDelete = (id) => {
    usersDelete(id);
  };

  const handleOnChange = (e) => {
    if (
      e.target.id == "street" ||
      e.target.id == "city" ||
      e.target.id == "zipcode"
    ) {
      setUserData({
        ...userData,
        address: { ...userData.address, [e.target.id]: e.target.value },
      });
    } else {
      setUserData({ ...userData, [e.target.id]: e.target.value });
    }
  };

  return (
    <div>
      <Dialog open={open} onClose={handleClose}>
        <DialogTitle>Update User Details</DialogTitle>
        <DialogContent>
          <TextField
            autoFocus
            margin="dense"
            id="name"
            label="Name"
            type="text"
            fullWidth
            onChange={(e) => handleOnChange(e)}
            variant="standard"
            value={userData.name}
          />
          <TextField
            autoFocus
            margin="dense"
            id="username"
            label="UserName"
            type="text"
            fullWidth
            onChange={(e) => handleOnChange(e)}
            variant="standard"
            value={userData.username}
          />
          <TextField
            autoFocus
            margin="dense"
            id="email"
            label="Email"
            type="email"
            fullWidth
            onChange={(e) => handleOnChange(e)}
            variant="standard"
            value={userData.email}
          />
          <Grid container spacing={2}>
            <Grid item xs={4}>
              <TextField
                autoFocus
                margin="dense"
                id="street"
                label="street"
                type="text"
                fullWidth
                onChange={(e) => handleOnChange(e)}
                variant="standard"
                value={userData && userData.address && userData.address.street}
              />
            </Grid>
            <Grid item xs={4}>
              <TextField
                autoFocus
                margin="dense"
                id="city"
                label="City"
                type="text"
                fullWidth
                onChange={(e) => handleOnChange(e)}
                variant="standard"
                value={userData && userData.address && userData.address.city}
              />
            </Grid>
            <Grid item xs={4}>
              <TextField
                autoFocus
                margin="dense"
                id="zipcode"
                label="Zipcode"
                type="text"
                fullWidth
                onChange={(e) => handleOnChange(e)}
                variant="standard"
                value={userData && userData.address && userData.address.zipcode}
              />
            </Grid>
          </Grid>
        </DialogContent>
        <DialogActions>
          <Button id='updateBtn' onClick={() => handleUpdate()}>Update</Button>
        </DialogActions>
      </Dialog>

      <TableContainer component={Paper}>
        <Table sx={{ minWidth: 650 }} aria-label="simple table">
          <TableHead>
            <TableRow>
              <TableCell>Id</TableCell>
              <TableCell>Name</TableCell>
              <TableCell>UserName</TableCell>
              <TableCell>Email</TableCell>
              <TableCell>Address</TableCell>
              <TableCell>Edit/Delete</TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            {userList &&
              userList.length > 0 &&
              userList.map((item) => (
                <TableRow
                  key={item.id}
                  sx={{ "&:last-child td, &:last-child th": { border: 0 } }}
                >
                  <TableCell component="th" scope="row">
                    {item.id}
                  </TableCell>
                  <TableCell>{item.name}</TableCell>
                  <TableCell>{item.username}</TableCell>
                  <TableCell>{item.email}</TableCell>
                  <TableCell>
                    {item.address.street +
                      "," +
                      item.address.city +
                      "," +
                      item.address.zipcode}
                  </TableCell>
                  <TableCell>
                    <Button id='editBtn' onClick={() => handleEdit(item.id)}>Edit</Button>
                    <Button id='deleteBtn' onClick={() => handleDelete(item.id)}>
                      Delete
                    </Button>
                  </TableCell>
                </TableRow>
              ))}
          </TableBody>
        </Table>
      </TableContainer>
    </div>
  );
}

const mapStateToProps = (state) => {
  return {
    Data: state.usersList,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    usersList: () => dispatch(usersList()),
    usersDelete: (id) => dispatch(usersDelete(id)),
    usersUpdate: (userData) => dispatch(usersUpdate(userData)),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(UserDetails);
