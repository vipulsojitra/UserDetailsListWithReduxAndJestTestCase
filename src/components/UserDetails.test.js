// import renderer from "react-test-renderer";
import { Button, Dialog, TextField } from "@mui/material";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { UserDetails } from "./UserDetails";

Enzyme.configure({ adapter: new Adapter() });
let data = [
  {
    id: 1,
    name: "Leanne Graham",
    username: "Bret",
    email: "Sincere@april.biz",
    address: {
      street: "Kulas Light",
      city: "Gwenborough",
      zipcode: "92998-3874",
    },
  },
];
const initialProps = {
  usersList: jest.fn(),
  usersDelete: jest.fn(),
  usersUpdate: jest.fn(),
  Data: { userList: data },
};
describe("UserDetails", () => {
  it("Component Render", () => {
    const component = shallow(<UserDetails {...initialProps} />);

    expect(component).toMatchSnapshot();
  });

  it("Check handleUpdate", () => {
    const component = shallow(<UserDetails {...initialProps} />);

    expect(component.find(Button)).toHaveLength(3);
    component.find("#updateBtn").props().onClick();
    expect(initialProps.usersUpdate).toHaveBeenCalled();
    expect(component.find(Dialog).props().open).toEqual(false);
  });
  it("Check handleEdit", () => {
    const component = shallow(<UserDetails {...initialProps} />);

    expect(component.find(Button)).toHaveLength(3);
    component.find("#editBtn").props().onClick();
    expect(component.find("#name").props().value).toEqual(data[0].name);
  });
  it("Check handleDelete", () => {
    const component = shallow(<UserDetails {...initialProps} />);

    expect(component.find(Button)).toHaveLength(3);
    component.find("#deleteBtn").props().onClick();
    expect(initialProps.usersDelete).toHaveBeenCalled();
  });

  it("Check onChange for name", () => {
    const component = shallow(<UserDetails {...initialProps} />);

    expect(component.find(TextField)).toHaveLength(6);

    let e = {
      target: {
        id: "name",
        value: "abc",
      },
    };
    component.find("#name").props().onChange(e);
    expect(component.find("#name").props().value).toEqual("abc");
  });

  it("Check onChange for username", () => {
    const component = shallow(<UserDetails {...initialProps} />);

    expect(component.find(TextField)).toHaveLength(6);

    let e = {
      target: {
        id: "username",
        value: "abc",
      },
    };

    component.find("#username").props().onChange(e);
    expect(component.find("#username").props().value).toEqual("abc");
  });

  it("Check onChange for email", () => {
    const component = shallow(<UserDetails {...initialProps} />);

    expect(component.find(TextField)).toHaveLength(6);

    let e = {
      target: {
        id: "email",
        value: "abc",
      },
    };

    component.find("#email").props().onChange(e);
    expect(component.find("#email").props().value).toEqual("abc");
  });

  it("Check onChange for city", () => {
    const component = shallow(<UserDetails {...initialProps} />);

    expect(component.find(TextField)).toHaveLength(6);

    let e = {
      target: {
        id: "city",
        value: "abc",
      },
    };

    component.find("#city").props().onChange(e);
    expect(component.find("#city").props().value).toEqual("abc");
  });

  it("Check onChange for zipcode", () => {
    const component = shallow(<UserDetails {...initialProps} />);

    expect(component.find(TextField)).toHaveLength(6);

    let e = {
      target: {
        id: "zipcode",
        value: "123",
      },
    };

    component.find("#zipcode").props().onChange(e);
    expect(component.find("#zipcode").props().value).toEqual("123");
  });

  it("Check onChange for street", () => {
    const component = shallow(<UserDetails {...initialProps} />);

    expect(component.find(TextField)).toHaveLength(6);

    let e = {
      target: {
        id: "street",
        value: "abc",
      },
    };
    component.find("#street").props().onChange(e);
    expect(component.find("#street").props().value).toEqual("abc");
  });

  it("Check modal onClose event", () => {
    const component = shallow(<UserDetails {...initialProps} />);

    component.find(Dialog).props().onClose();
    expect(component.find(Dialog).props().open).toEqual(false);
  });
});
