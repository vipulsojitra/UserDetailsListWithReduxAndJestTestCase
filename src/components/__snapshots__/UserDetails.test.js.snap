// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDetails Check handleDelete 1`] = `ShallowWrapper {}`;

exports[`UserDetails Check handleEdit 1`] = `ShallowWrapper {}`;

exports[`UserDetails Check handleUpdate 1`] = `ShallowWrapper {}`;

exports[`UserDetails Check modal onClose event 1`] = `ShallowWrapper {}`;

exports[`UserDetails Check onChange for city 1`] = `ShallowWrapper {}`;

exports[`UserDetails Check onChange for email 1`] = `ShallowWrapper {}`;

exports[`UserDetails Check onChange for name 1`] = `ShallowWrapper {}`;

exports[`UserDetails Check onChange for street 1`] = `ShallowWrapper {}`;

exports[`UserDetails Check onChange for username 1`] = `ShallowWrapper {}`;

exports[`UserDetails Check onChange for zipcode 1`] = `ShallowWrapper {}`;

exports[`UserDetails Component Render 1`] = `ShallowWrapper {}`;
